### PR TITLE
When calling Nodes#setNodes, NodeListener methods should be called as required

### PR DESCRIPTION
**Pull request by Vlatombe**
For example, when loading a new casc configuration, it calls `Nodes#setNodes`, and the `NodeListener` methods were not called, causing listener implementations not to be notified according to the new state.

### Testing done

Automated tests pass.

### Proposed changelog entries

- Calls to `Nodes#setNodes` now notify `NodeListener` implementations.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [X] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [X] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/jenkins/model/Nodes.java
Comment (jglick) [inline core/src/main/java/jenkins/model/Nodes.java:123]: Perhaps simpler to use a `Map<Sttring, Node>` for `toRemove`.

File: test/src/test/java/jenkins/model/NodesTest.java
Comment (jglick) [inline test/src/test/java/jenkins/model/NodesTest.java:253]: Test would be clearer using string constants rather than `Node.toString` I think, e.g.
```suggestion
        assertThat(CheckSetNodes.created, containsInAnyOrder("nodeA", "nodeB"));
```

File: test/src/test/java/jenkins/model/NodesTest.java
Comment (jglick) [inline test/src/test/java/jenkins/model/NodesTest.java:273]: `Node.getNodeName` seems a more stable choice for testing.

File: test/src/test/java/jenkins/model/NodesTest.java
Comment (jglick) [inline test/src/test/java/jenkins/model/NodesTest.java:258]: A little weak since this is not asserting the other collections in each case. Better after each `setNodes` call to assert the contents of all three collections. (Fancier test utilities have a method to return the set of recent notifications but also clear it, so the next call will be an empty list unless there is actually a _new_ notification there.)

**Review by jglick — APPROVED** [review]

**Review by jglick — APPROVED** [review]

**Review by timja — APPROVED** [review]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
